### ISSUE TITLE
Missing option to verify on Save

### ISSUE THREAD

**Issue by franck44** [DafnyVSCode/ide-vscode#18]
Thanks for updating the Vscode Dafny plugin.

There used to be a configuration  option to "verify on save" rather than continuously.
Has it been removed?

After quite a few attempts, I did not manage to successfully run the new plugin.
It seems it verifies only one file in the list of files opened and it is unclear which/when it does so.
Moreover, It seems like the "problems" tab in VSCode does not get any refresh and displays the verification results only once.

I am experiencing these problems on MacBookPro, running Big Sur.
The DafnyLS.dll (and other files) are correctly installed in my .vscode/extensions/correctnesslab.dafny-vscode-1.3.0
and if I run them manually (using botnet) they work as expected.

**Comment by camrein** [comment]
The language server has been reworked from the ground up. Therefore, it may lack a few features. Instead of nitpicking which features might be relevant for end-users, we chose the path to release a new version and wait for feedback and decide upon them.
To clarify your question: Verification happens every time you change or save your active document. It may well be that your code requires unexpectedly long to verify - giving you the feel nothing happens at all. You could check my assumption the following way:

- How long does the Dafny compiler require to compile your program?
- Changing the code to contain any syntactic or semantic error. In contrast to verification, checking the syntax and semantic model should happen instantly.
- Create a small test case such as a method containing `assert false`.

Finally, I have to note the following shortcoming of the language server: Only one document may be parsed, semantically checked, and verified at a time because some of the components (i.e., the Boogie and Dafny compilers) are not thread-safe. Having a document that requires a long time to verify will limit the overall experience since other documents have to wait.

**Comment by franck44** [comment]
Thanks for the feedback.
> Only one document may be parsed, semantically checked, and verified at a time because some of the components (i.e., the Boogie and Dafny compilers) are not thread-safe. 

Yes I understand that but it does not mean we cannot perform sequentially the analyses of all the open files, correct?

> To clarify your question: Verification happens every time you change or save your active document. It may well be that your code requires unexpectedly long to verify 

The drawback is that at almost every character a Dafny verification (and z3) is started. If a given file takes a long time to verify, starting a verification makes it even worse.
Software verification with a tool like Dafny is inherently costly (time-wise) so it may be better to control when we want the verification to start (e.g. on save).
And even with a lag of a few seconds (which is not unexpectedly long) it is hard to use.

At the moment the plugin is unusable for me in all the examples I have: too much lag to get results, the results when they are displayed (Problems tab) correspond to a snapshot a few seconds ago when I was typing something. 

As a result I have reverted to the old (legacy) plugin and configured it as I did before with dafny3-pre-release2.

**Comment by camrein** [comment]
You are right. In fact, the VSCode should actually queue the analysis requests. I can't tell what happens if there are exceptionally many pending requests. What I meant is that the current design may give the feeling that nothing happens at all. Moreover, analysis results are discarded when closing a document - this could further lead to incomprehensible verification behavior.

Nonetheless, I agree that working with a more or less unresponsive IDE is a pain. Therefore, I have extended the language server so you can configure the verification to your desire. Since the final release of Dafny 3.0.0 is around the corner, I will delay the update until then.

**Comment by camrein** [comment]
Since Dafny 3.0.0 is now officially released, I have updated both the language server and the VSCode plugin. I hope the latest changes make the plugin more beneficial for your use.

**Comment by franck44** [comment]
@camrein Thanks a lot for your help (and for maintaining the plugin!)

**Comment by camrein** [comment]
I am glad I could help you. Please do not hesitate to open another issue if you have any other inquiries.
